interfaces/builtin/dbus_test.go

```diff
@@ -256,6 +299,200 @@ slots:
+  app-dbus-slot:
+    daemon: simple
+    daemon-scope: user
+    activates-on:
+    - dbus-service-slot
+  app-dbus-slot-duplicated:
+    daemon: simple
+    daemon-scope: user
+    activates-on:
+    - dbus-service-slot
+`
+	info := snaptest.MockInfo(c, mockSnapYaml, nil)
+	slot := info.Slots["dbus-service-slot"]
+	err := interfaces.BeforePrepareSlot(s.iface, slot)
+	c.Assert(err, ErrorMatches, `cannot add activatable dbus service slot to multiple apps`)
+}
+
+func (s *DbusInterfaceSuite) TestSanitizeSlotActivatableWrongBus(c *C) {
+	var mockSnapYaml = `name: dbus-snap
+version: 1.0
+slots:
+  dbus-session-service-slot:
+    interface: dbus
+    bus: session
+    name: org.dbus-snap.session
+  dbus-system-service-slot:
+    interface: dbus
+    bus: system
+    name: org.dbus-snap.session
+apps:
+  system-service:
+    daemon: simple
+    slots:
+    - dbus-session-service-slot
+    activates-on:
+    - dbus-session-service-slot
+  session-service:
+    daemon: simple
+    daemon-scope: user
+    slots:
+    - dbus-system-service-slot
+    activates-on:
+    - dbus-system-service-slot
+`
+	info := snaptest.MockInfo(c, mockSnapYaml, nil)
+	slot := info.Slots["dbus-session-service-slot"]
+	err := interfaces.BeforePrepareSlot(s.iface, slot)
+	c.Check(err, ErrorMatches, `system daemons can only activate from the D-Bus system bus`)
+
+	slot = info.Slots["dbus-system-service-slot"]
+	err = interfaces.BeforePrepareSlot(s.iface, slot)
+	c.Check(err, ErrorMatches, `user daemons can only activate from the D-Bus session bus`)
+}
+
+func (s *DbusInterfaceSuite) TestSanitizeSlotActivatableSessionConflict(c *C) {
+	var mockSnapYaml = `name: dbus-snap
+version: 1.0
+slots:
+  dbus-service-slot:
+    interface: dbus
+    bus: session
+    name: org.dbus-snap.session
+apps:
+  app-dbus-slot:
+    daemon: simple
+    daemon-scope: user
+    activates-on:
+    - dbus-service-slot
+`
+
+	err := os.MkdirAll(dirs.SnapDBusSessionServicesDir, 0755)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(dirs.SnapDBusSessionServicesDir, "org.dbus-snap.session.service"), []byte(`[D-BUS Service]
+Name=org.dbus-snap.session
+Exec=command
+X-Snap=other-snap
+`), 0644)
+	c.Assert(err, IsNil)
+
+	info := snaptest.MockInfo(c, mockSnapYaml, nil)
+	slot := info.Slots["dbus-service-slot"]
+	err = interfaces.BeforePrepareSlot(s.iface, slot)
+	c.Assert(err, ErrorMatches, `bus name "org.dbus-snap.session" is already owned by snap "other-snap"`)
+}
+
+func (s *DbusInterfaceSuite) TestSanitizeSlotActivatableSystemConflict(c *C) {
+	var mockSnapYaml = `name: dbus-snap
+version: 1.0
+slots:
+  dbus-service-slot:
+    interface: dbus
+    bus: system
+    name: org.dbus-snap.system
+apps:
+  app-dbus-slot:
+    daemon: simple
+    activates-on:
+    - dbus-service-slot
+`
+
+	err := os.MkdirAll(dirs.SnapDBusSystemServicesDir, 0755)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(dirs.SnapDBusSystemServicesDir, "org.dbus-snap.system.service"), []byte(`[D-BUS Service]
```

Comment:
Hmm, this makes me somewhat worried we are duplicating state facilities. Instead of looking at the disk snapd should determine this via the state. This also frees us from parsing D-Bus service files as a authoritative source of information.

This also probably suggests that bus names should be something snapd models with regards to name clashes, similar to aliases and eventually man pages. I'm not sure how to structure this yet, though.